### PR TITLE
Document testing strategy and patterns.

### PR DESCRIPTION
**Pull request by fisx**

### CODE REVIEW

**Comment by fisx** [comment]
@mythsunwind o/

**Review by mythsunwind — APPROVED** [review]

File: docs/developer/testing.md [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+All pure functions `f` that do something interesting should have a couple of tests of the form `shouldBe (f <args>) <result>` to cover corner cases.
+If code is refactored, all the effects should be mocked with polysemy or mtl, the old implementation should be moved to the test suite, and there should be quickcheck tests running the new implementation against the old and comparing results ([example](https://github.com/wireapp/wire-server/blob/develop/services/gundeck/test/unit/MockGundeck.hs)).
+
+### Integration tests
+
+- All new rest API end-points need to be called a few times to cover as much of the corner cases as feasible.
+- We have machinery to set up schaffolding teams and modify context such as server configuration files where needed.  Look through the existing code for inspiration.
```
Comment (pcapriotti) [inline docs/developer/testing.md:24]: Typo: "schaffolding".

**Review by pcapriotti — APPROVED** [review]
Looks good to me!
